crates/wasi-provider/src/states/starting.rs

```diff
@@ -0,0 +1,136 @@
+        volume_mounts
+            .iter()
+            .map(|vm| -> anyhow::Result<(PathBuf, Option<PathBuf>)> {
+                // Check the volume exists first
+                let vol = volumes.get(&vm.name).ok_or_else(|| {
+                    anyhow::anyhow!(
+                        "no volume with the name of {} found for container {}",
+                        vm.name,
+                        container.name()
+                    )
+                })?;
+                let mut guest_path = PathBuf::from(&vm.mount_path);
+                if let Some(sub_path) = &vm.sub_path {
+                    guest_path.push(sub_path);
+                }
+                // We can safely assume that this should be valid UTF-8 because it would have
+                // been validated by the k8s API
+                Ok((vol.deref().clone(), Some(guest_path)))
+            })
+            .collect::<anyhow::Result<HashMap<PathBuf, Option<PathBuf>>>>()
+    } else {
+        Ok(HashMap::default())
+    }
+}
+
+async fn run_init_container(
+    pod_state: &mut PodState,
+    pod: &Pod,
+    container: &Container,
+) -> anyhow::Result<kubelet::container::Handle<wasi_runtime::Runtime, wasi_runtime::HandleFactory>>
+{
+    info!(
+        "Running init container {:?} for pod {:?}",
+        container.name(),
+        pod.name()
+    );
+    let mut handle = start_container(pod_state, pod, &container).await?;
+    // TODO(bacongobbler): add some form of a timeout here in case an init container never finishes
+    handle.wait().await?;
+    Ok(handle)
+}
+
+async fn start_container(
+    pod_state: &mut PodState,
+    pod: &Pod,
+    container: &Container,
+) -> anyhow::Result<kubelet::container::Handle<wasi_runtime::Runtime, wasi_runtime::HandleFactory>>
+{
+    let module_data = pod_state
+        .run_context
+        .modules
+        .remove(container.name())
+        .expect("FATAL ERROR: module map not properly populated");
+    let client = kube::Client::new(pod_state.shared.kubeconfig.clone());
+    let env = provider::env_vars(&container, pod, &client).await;
+    let args = container.args().clone().unwrap_or_default();
+    let volumes = Ref::volumes_from_pod(&pod_state.shared.volume_path, &pod, &client).await?;
+    let container_volumes = volume_path_map(container, &volumes)?;
+
+    let runtime = WasiRuntime::new(
+        container.name().to_owned(),
+        module_data,
+        env,
+        args,
+        container_volumes,
+        pod_state.shared.log_path.clone(),
+        pod_state.run_context.status_sender.clone(),
+    )
+    .await?;
+
+    debug!("Starting container {} on thread", container.name());
+    runtime.start().await
+}
+
+state!(
+    /// The Kubelet is starting the Pod.
+    Starting,
+    PodState,
+    {
+        let mut container_handles = HashMap::new();
+
+        info!("Running init containers for pod {:?}", pod.name());
+        for init_container in pod.init_containers() {
```

Comment:
I patched the `init_containers` method, it was returning the Pod's app containers. Was this intentional? How was it passing tests before?

Reply:
No, I don't think that was intentional. Thank you for catching that.

I see your change in https://github.com/bacongobbler/krustlet/commit/02d15a8e6104b885a6704c4f8b4eec0722cbe36f to `all_containers` accidentally iterating through `self.containers` twice, but I don't see a fix for `pod.init_containers()`. Can you point out the patch for `init_containers()`?

Reply:
Ah, I misremembered what I patched, it sounds like `all_containers` was the only method impacted.